Robus/src/msgAlloc.c

```diff
@@ -17,45 +18,134 @@
  * Functions
  ******************************************************************************/
 
+/*******************************************************************************
+ * Functions --> Allocator memory manager
+ ******************************************************************************/
+
+static void MsgAlloc_ClearSlot(uint16_t _alloc_id);
+static void MsgAlloc_FindAllocSlot(void);
+static error_return_t MsgAlloc_ClearMemSpace(void *from, void *to);
+
+/******************************************************************************
+ * @brief Clear a slot. This action is due to an error
+ * @param None
+ * @return None
+ ******************************************************************************/
+static void MsgAlloc_ClearSlot(uint16_t _alloc_id)
+{
+    // TODO we should set an error into the concerned module
+    alloc_table[_alloc_id].msg_pt = 0;
+    alloc_table[_alloc_id].vm_pt = 0;
+}
+
 /******************************************************************************
  * @brief get the next free alloc_table slot.
  * @param None
  * @return None
  ******************************************************************************/
 static void MsgAlloc_FindAllocSlot(void)
 {
-    /******** Find the next available slotslot **********
-     * - loop from alloc_id to MAX_MSG_NB to find a free slot set alloc_id and return if there is one
-     * - loop from 0 to alloc_id to find a free slot set alloc_id and return if there is one
-     * - At this point there is no slot available, put an error somewhere
-     * - remove the message on the next slot and make it clean for the module of this slot
-     * - set alloc_id to the next slot pointer and return .
+    /* In this function we search for an available slot. 
+     * If a module doesn't consume it's message for a while we can jump some slot to find an available one.
+     * But this message that we jump is now considered as one of the newest.
+     * To keep slots ordered by time we should move this slot 
      */
+    //******** check if the next slot is available **********
+    if ((alloc_id < MAX_MSG_NB - 1) & (alloc_table[alloc_id + 1].msg_pt == 0))
+    {
+        alloc_id++;
+        return;
+    }
+    if ((alloc_id == MAX_MSG_NB - 1) & (alloc_table[0].msg_pt == 0))
+    {
+        alloc_id = 0;
+        return;
+    }
+    // The next slot is not available we have to find one and slide older slots.
+
+    //******** Find the next available slotslot **********
+    // loop from alloc_id+1 to MAX_MSG_NB to find a free slot
+    //set alloc_id
+    //return if there is one
+    for (uint16_t i = alloc_id + 1; i < MAX_MSG_NB; i++)
+    {
+        if (alloc_table[i].msg_pt == 0)
+        {
+            // we found one
+            // slide older slots to let the next slot available
+            memcpy((void *)&alloc_table[alloc_id + 2], (void *)&alloc_table[alloc_id + 1], (alloc_id + 1 - i) * sizeof(msg_allocator_t));
```

Comment:
use memmove instead of memcpy but memove seams to do 2 copy. The only thing we need is to copy data by decrementing pointer instead of incementing if as memcpy do.
We should revers the memcpy algo to save some execution time : https://stackoverflow.com/questions/17591624/understanding-the-source-code-of-memcpy

Reply:
Based on comments we will remove the copy here and make memory movement on message pull.